File: problems/tier1-foundation/003-notification-system/tests/cpp/part2_test.cpp
// Part 2 Tests — Priority Filtering
// Tests that notifications respect per-user priority preferences

#include <cassert>
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

int part2_tests() {
    int passed = 0;
    int failed = 0;

    // Test 1: critical event reaches all users regardless of min priority
    try {
        User u1 = {"u1", "u1@test.com", "+1-555-0001", {"email"}};
        unordered_map<string, string> prefs = {{"*", "critical"}};
        // Should not throw
        notify("CRITICAL: System down", "critical", {u1}, prefs);
        cout << "PASS test_critical_reaches_all" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_critical_reaches_all" << endl;
        failed++;
    }

    // Test 2: promotional event blocked when user wants info+ only
    try {
        User u1 = {"u1", "u1@test.com", "+1-555-0001", {"email"}};
        unordered_map<string, string> prefs = {{"*", "info"}};
        // Promotional should be filtered out — no exception expected
        notify("50% off sale!", "promotional", {u1}, prefs);
        cout << "PASS test_promotional_filtered" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_promotional_filtered" << endl;
        failed++;
    }

    // Test 3: empty priority prefs defaults to sending all events
    try {
        User u1 = {"u1", "u1@test.com", "+1-555-0001", {"email"}};
        unordered_map<string, string> emptyPrefs;
        notify("Informational update", "info", {u1}, emptyPrefs);
        cout << "PASS test_empty_prefs_allow_all" << endl;
        passed++;
    } catch (...) {
        cout << "FAIL test_empty_prefs_allow_all" << endl;
        failed++;
    }

    cout << "PART2_SUMMARY " << passed << "/" << (passed + failed) << endl;
    return failed;
}

File: problems/tier1-foundation/003-notification-system/tests/java/Part2Test.java
// Notification System — Part 2 Tests
import java.util.*;
import java.util.stream.*;

class Part2Test {
    static boolean testCriticalReachesAll() {
        try {
            User u1 = Arrays.asList("u1", "u1@test.com", "+1-555-0001", {"email")};
            Map<String, String> prefs = {Arrays.asList("*", "critical")};
            // Should not throw
            notify("CRITICAL: System down", "critical", {u1}, prefs);
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testCriticalReachesAll");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testCriticalReachesAll (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testPromotionalFiltered() {
        try {
            User u1 = Arrays.asList("u1", "u1@test.com", "+1-555-0001", {"email")};
            Map<String, String> prefs = {Arrays.asList("*", "info")};
            // Promotional should be filtered out — no exception expected
            notify("50% off sale!", "promotional", {u1}, prefs);
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testPromotionalFiltered");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testPromotionalFiltered (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    static boolean testEmptyPrefsAllowAll() {
        try {
            User u1 = Arrays.asList("u1", "u1@test.com", "+1-555-0001", {"email")};
            Map<String, String> emptyPrefs;
            notify("Informational update", "info", {u1}, emptyPrefs);
            boolean pass = true; // FIXME: no assertions found
            System.out.println((pass ? "PASS" : "FAIL") + ": testEmptyPrefsAllowAll");
            return pass;
        } catch (Exception e) {
            System.out.println("FAIL: testEmptyPrefsAllowAll (exception: " + e.getMessage() + ")");
            return false;
        }
    }

    public static int runTests() {
        int passed = 0, total = 0;
        total++; if (testCriticalReachesAll()) passed++;
        total++; if (testPromotionalFiltered()) passed++;
        total++; if (testEmptyPrefsAllowAll()) passed++;
        System.out.println("PART2_SUMMARY " + passed + "/" + total);
        return passed;
    }
}
